### PR TITLE
Add sse support

### LINKED ISSUE #503: Important delay with streams (Mozilla EventSource)
I'm exposing through Traefik an [home-assistant](https://github.com/home-assistant/home-assistant/) instance. It's using [Mozilla's EventSource](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events) to send events back to the front-end.

If I'm accessing directly the home-assistant instance, events are nearly instant. If I access it through Traefik, they are sent, but with an important delay (between 30 seconds and a minute or two). It looks like a buffer flush occurs every x seconds: if I send a lot of events, they are sent quickly. If I only have one or two, they are queued for a while.

It may be a feature, but not documented anywhere (or at least I couldn't find it). If it is, is there any way to configure the behavior?

### LINKED ISSUE #524: Error with -consulcatalog and missing load balance method on 1.0.0
Running the docker image with tag v1.0.0 against a consul catalog results in the following message being sent to the console for every service in consul repeatedly: 

`time="2016-07-13T03:06:13Z" level=warning msg="Error loading load balancer method '<nil>' for backend backend-service-name: Invalid method, using default. Using default wrr."`

Downgrading to v1.0.0-rc3 fixes the issue.

### PR DESCRIPTION
**Pull request by emilevauge**
This PR adds [SSE](https://en.wikipedia.org/wiki/Server-sent_events) support.
Fixes #503 
Fixes #524

### CODE REVIEW

**Comment by JayH5** [comment]
Hey @emilevauge thanks for all your work on Traefik. It's been really useful for us.

Looking at containous/oxy@98f8b7d it looks like your approach here is to treat the `text/event-stream` content-type as a special case and then reading each event, writing each event to the `http.ResponseWriter` and then flushing it.

First of all, it looks like you are splitting events on `\n\n` which is a problem because the [Server-Sent Events spec](https://html.spec.whatwg.org/multipage/comms.html#server-sent-events) also allows `\r` or `\r\n` to delimit lines, not just `\n`.

But what I really want to ask is if maybe there is a more general solution to this problem?

The underlying problem seems to be that Go's `http.ResponseWriter` buffers responses by default which breaks streaming protocols (as far as I understand -- I am very far from being a golang expert). Wouldn't it make more sense to have an option to disable that buffering behaviour and immediately flush all request body content to the response? Let the client worry about parsing the SSE content itself, not Traefik.

As far as I can see, there are 2 ways that buffering could be disabled:
 1. An option in the frontend, say, something like a `flush` boolean flag. It seems like this could be passed through to `oxy/forward/fwd` the same way `passHostHeaders` is.
 2. A special header to indicate that a response should not be buffered. I'm not sure about other software, but Nginx has the [`X-Accel-Buffering`](https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/#x-accel-buffering) header that can be used to switch off buffering. Having a special header for this means that a) it can be done on a per-response basis which is useful, and b) it can be done for any streaming content type, not just a specific one.

Anyway, just my thoughts. Thanks again for the great project.

**Comment by emilevauge** [comment]
Hi @JayH5 
Ops, I read the SSE specs to quickly...
I absolutely agree with you on using a more generic way of dealing with this issue. That's what I wanted to do at the beginning, but this is not that simple.

1. In the std lib, in the reverse proxy implementation, https://golang.org/src/net/http/httputil/reverseproxy.go, they give the possibility to flush the buffer at each `FlushInterval`. This helps but you have to configure the interval accordingly to your needs...
Here, the issue is not in `http.ResponseWriter`, but in how we copy the bytes in `fwd.serveHTTP` in oxy. We could also flush the response at each byte received (without waiting for a `\n\n` or `\r\n`) but I have to test the performance of that solution.
2. Once the technical solution will be found on how to NOT buffer responses, we will have to decide how we activate this. What I like in using `text/event-stream` is that it's automatic, no need for an option to be set by the user. But we could add an option to force enable this in any case, in frontend section maybe?

Any help on both 1/ and 2/ is welcomed :)

/cc @containous/traefik

**Comment by Russell-IO** [comment]
LGTM

**Comment by JayH5** [comment]
@emilevauge: Thanks for the response.

I wasn't aware of the golang stdlib reverse proxy (again, really not a golang expert). Seems strange to me to have a timer-based flushing system 😕. But there's some code in there that nearly gets at what I think could work.

I wasn't suggesting writing and flushing data byte-by-byte. But it should be possible to read as much as is available into a buffer and then write and flush.

From what [I can see](https://github.com/vulcand/oxy/blob/40720199a16cdfb3854ff58ad91d290a7dd56de0/forward/fwd.go#L189), oxy uses `io.Copy` to copy the response body, which in turn calls [`io.CopyBuffer`](https://golang.org/src/io/io.go?s=12575:12653#L347). That method copies over (up to) 32KB chunks of data from the Reader to the Writer (if you don't give it your own buffer).

It looks like `io.Reader` has the behaviour we would want. From the docs: *"If some data is available but not len(p) bytes, Read conventionally returns what is available instead of waiting for more."*. So hopefully we can read as much as is available into a buffer then write and flush.

The simplest way I can see to do this would be to just flush the Writer after each buffer is written to it. Copying code from the stdlib reverse proxy it may be possible to do something like this, where the call oxy makes to `io.Copy` is replaced with `copyBody`?
```go
func (f *httpForwarder) copyBody(dst io.Writer, src io.Reader) {
	if f.flush { // Or check for header
		if wf, ok := dst.(writeFlusher); ok {
			fw := &flushWriter{
				dst: wf,
			}
			dst = fw
		}
	}

	return io.Copy(dst, src)
}

type writeFlusher interface {
	io.Writer
	http.Flusher
}

type flushWriter struct {
	dst writeFlusher
}

func (m *flushWriter) Write(p []byte) (int, error) {
	written, err := m.dst.Write(p)
	m.dst.Flush()
	return written, err
}
```

(This is the most golang I've ever written so there are probably issues with the above).

**Comment by emilevauge** [comment]
@JayH5: I made some changes in oxy to stream the body without analyzing it: https://github.com/containous/oxy/commit/ca6b2dd809472073cdf75199f3bb3c9f1b5a90a9
I just modified the code from `copyBuffer`: https://golang.org/src/io/io.go?s=12575:12653#L366 to flush at each iteration.
I will add later (in 1.1) an option to force streaming in frontend but in 1.0.1, I will rely on `Content-Type` header.
WDYT?

File: server.go [hunk 1]
```diff
@@ -97,20 +98,37 @@ func (server *Server) Start() {
 
 // Stop stops the server
 func (server *Server) Stop() {
-	for _, serverEntryPoint := range server.serverEntryPoints {
-		serverEntryPoint.httpServer.BlockingClose()
+	for serverEntryPointName, serverEntryPoint := range server.serverEntryPoints {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Duration(server.globalConfiguration.GraceTimeOut)*time.Second)
+		go func(ctx context.Context) {
```
Comment (vdemeester) [inline server.go]: Hum, why passing `ctx` if you don't use it ? 👼

**Comment by vdemeester** [comment]
> I will add later (in 1.1) an option to force streaming in frontend but in 1.0.1, I will rely on `Content-Type` header.

The less changes for 1.0.1, the better :wink:

**Comment by JayH5** [comment]
@emilevauge: I've done a bit more digging in the go source code.. and .. it's kind of tricky. I don't think containous/oxy@d12fc74 is going to work. `http.ResponseWriter` will usually implement `io.ReaderFrom`, as it does in the golang source [here](https://golang.org/src/net/http/server.go?s=1893:2975#L418), so the flush call you added will never be reached. It's doubly complicated because when `ReadFrom` is called on the `http.ResponseWriter`, it calls `io.CopyBuffer` anyway while hiding it's `io.ReaderFrom` interface so that a regular copy is done 😧 

I would still suggest my route of wrapping all calls to `Write()` with a following `Flush()`. This whole thing really needs tests I guess. I could maybe make a PR to `containous/oxy`?

**Comment by emilevauge** [comment]
@JayH5 :
>I don't think containous/oxy@d12fc74 is going to work

The previous solution was working well in my tests ;) But I agree we could make it better using a wrapper.

Here is what I propose. I created a wrapper `responseFlusher`. This wrapper must implement `http.Hijacker`, `http.Flusher` and `http.CloseNotifier` to stay compliant with all possible response writers:

```
package forward

import (
	"bufio"
	"fmt"
	"net"
	"net/http"
)

var (
	_ http.Hijacker      = &responseFlusher{}
	_ http.Flusher       = &responseFlusher{}
	_ http.CloseNotifier = &responseFlusher{}
)

type responseFlusher struct {
	http.ResponseWriter
	flush bool
}

func newResponseFlusher(rw http.ResponseWriter, flush bool) *responseFlusher {
	return &responseFlusher{
		ResponseWriter: rw,
		flush:          flush,
	}
}

func (wf *responseFlusher) Write(p []byte) (int, error) {
	written, err := wf.ResponseWriter.Write(p)
	if wf.flush {
		wf.Flush()
	}
	return written, err
}

func (wf *responseFlusher) Hijack() (net.Conn, *bufio.ReadWriter, error) {
	hijacker, ok := wf.ResponseWriter.(http.Hijacker)
	if !ok {
		return nil, nil, fmt.Errorf("the ResponseWriter doesn't support the Hijacker interface")
	}
	return hijacker.Hijack()
}

func (wf *responseFlusher) CloseNotify() <-chan bool {
	return wf.ResponseWriter.(http.CloseNotifier).CloseNotify()
}

func (wf *responseFlusher) Flush() {
	flusher, ok := wf.ResponseWriter.(http.Flusher)
	if ok {
		flusher.Flush()
	}
}
```

The `httpForwarder.serverHTTP` then calls it:
```
	stream := "text/event-stream" == response.Header.Get(ContentType) || f.streamResponse
	written, err := io.Copy(newResponseFlusher(w, stream), response.Body)
```

I will add a unit test in oxy for that use case.

BUT, some work will have to be done on this project because the existing tests are not all passing :'( I forked it since the original project is not really maintained anymore but I didn't add time to correct each existing test yet. If you can give some help on this, it would be really appreciated ;)

WDYT?

**Comment by emilevauge** [comment]
I updated `containous/oxy` https://github.com/containous/oxy/commit/ab7796d7036b425fbc945853cd1b7e7adf43b0d6.

**Comment by JayH5** [comment]
@emilevauge sorry I meant to reply but then got distracted trying to get the oxy tests to work. I tried but I think I'm going to give up.. all those chunked-encoding tests are broken and they're not straightforward to fix. It might almost be easier just to pull out the bits of oxy that Traefik uses and start from scratch.

The flushing code is looking better. There are a couple of places where I think it could be cleaned up a bit but otherwise it LGTM. Thanks.

**Comment by emilevauge** [comment]
@jayH5 thanks for your help in this :)

**Comment by vdemeester** [comment]
LGTM 👼

**Comment by Russell-IO** [comment]
LGTM

**Comment by ashmckenzie** [comment]
❤️

**Comment by JayH5** [comment]
@emilevauge: in practice, I found the check for the `Content-Type` value to be a little too simplistic. I made a PR to fix that: https://github.com/containous/oxy/pull/9

**Pull request merged**